done/done.py

```diff
@@ -1,77 +1,110 @@
 from xblock.fragment import Fragment
 
+try:
+    from eventtracking import tracker
+except ImportError:
+    class tracker(object):  # pylint: disable=invalid-name
+        """
+        Define tracker if eventtracking cannot be imported. This is a
+        workaround so that the code works in both edx-platform and
+        XBlock workbench (the latter of which does not support event
+        emission). This should be replaced with XBlock's emit(), but
+        at present, emit() is broken.
+        """
+        def __init__(self):
+            """ Do nothing """
+            pass
+
+        @staticmethod
+        def emit(param1, param2):
+            """ In workbench, do nothing for event emission """
+            pass
+
+
+def resource_string(path):
+    """Handy helper for getting resources from our kit."""
+    data = pkg_resources.resource_string(__name__, path)
+    return data.decode("utf8")
+
 
 class DoneXBlock(XBlock):
     """
     Show a toggle which lets students mark things as done.
     """
 
-    # Fields are defined on the class.  You can access them in your code as
-    # self.<fieldname>.
     done = Boolean(
-           scope = Scope.user_state, 
-           help = "Is the student done?",
-           default = False
-        )
+        scope=Scope.user_state,
+        help="Is the student done?",
+        default=False
+    )
 
     align = String(
-           scope = Scope.settings, 
-           help = "Align left/right/center",
-           default = "left"
-        )
+        scope=Scope.settings,
+        help="Align left/right/center",
+        default="left"
+    )
 
     has_score = True
 
-    def resource_string(self, path):
-        """Handy helper for getting resources from our kit."""
-        data = pkg_resources.resource_string(__name__, path)
-        return data.decode("utf8")
-
+    # pylint: disable=unused-argument
     @XBlock.json_handler
     def toggle_button(self, data, suffix=''):
-        self.done = data['done']
-        if data['done']:
-            grade = 1
-        else:
-            grade = 0
-
-        self.runtime.publish(self, 'grade', {'value':grade, 'max_value': 1})
-        return {}
-
-
-    def student_view(self, context=None):
+        """
+        Ajax call when the button is clicked. Input is a JSON dictionary
+        with one boolean field: `done`. This will save this in the
+        XBlock field, and then issue an appropriate grade.
+        """
+        if 'done' in data:
+            self.done = data['done']
+            if data['done']:
+                grade = 1
+            else:
+                grade = 0
+            grade_event = {'value': grade, 'max_value': 1}
+            self.runtime.publish(self, 'grade', grade_event)
+            # This should move to self.runtime.publish, once that pipeline
+            # is finished for XBlocks.
+            tracker.emit("edx.done.toggle", {'done': self.done})        
```

Comment:
The logging does work. I agree tests would be helpful. Do you know off-hand some place where we test events? I'm actually not quite sure of the best design patterns. I did a little bit of grepping, but most of the test cases were a little convoluted. Otherwise, I'll keep looking myself.

I have not looked at recent event design guidelines, but generally in the olden days, we'd use the same name for both events (see http://edx.readthedocs.org/en/latest/internal_data_formats/tracking_logs.html#problem-check-browser). It's pretty clear from the logs which event is client and which is server (the formats are a bit different). It's helpful to redundancy like this for checking integrity when reviewing events. If e.g. numbers of events or results don't match up too well, you know there's an issue. You also get client events even if the server is giving errors. The server, on the other hand, prevents spoofed events and similar.

If the guidelines have changed, we could change this if it is important. We have some number of events in this format already. @mulby What are the guidelines here?